File: lclang/src/main/java/lclang/libs/lang/classes/numbers/NumberClass.java
package lclang.libs.lang.classes.numbers;

import lclang.libs.lang.classes.BoolClass;
import lclang.libs.lang.classes.CharClass;
import lclang.libs.lang.classes.LibraryClass;
import lclang.libs.lang.classes.StringClass;
import lclang.types.Type;

public class NumberClass extends LibraryClass {
    public static final String NAME = "number";
    public static final NumberClass INSTANCE = new NumberClass();
    public static final Type TYPE = INSTANCE.classType;
    public Number number;

    private NumberClass() {
        super(NAME);
    }

    public NumberClass(String name) {
        super(name);
        extendsClass = INSTANCE;
    }

    public NumberClass(String name, Number number) {
        this(name);
        this.number = number;

        extendsClass = INSTANCE;
        globals.put("toInt", method((caller, lcClasses) -> IntClass.get(number.intValue()),
                IntClass.TYPE));
        globals.put("toLong", method((caller, lcClasses) -> LongClass.get(number.longValue()),
                LongClass.TYPE));
        globals.put("toDouble", method((caller, lcClasses) -> DoubleClass.get(number.doubleValue()),
                DoubleClass.TYPE));
        globals.put("toChar", method((caller, lcClasses) -> CharClass.get((char) number.intValue()),
                CharClass.type));
        globals.put("toString", method((caller, lcClasses) -> StringClass.get(String.valueOf(number)),
                StringClass.type));
    }

    public NumberClass add(NumberClass another) {
        if (this instanceof DoubleClass || another instanceof DoubleClass)
            return DoubleClass.get(number.doubleValue() + another.number.doubleValue());
        else if (this instanceof LongClass || another instanceof LongClass)
            return LongClass.get(number.longValue() + another.number.longValue());

        return IntClass.get(number.intValue() + another.number.intValue());
    }

    public NumberClass minus(NumberClass another) {
        if (this instanceof DoubleClass || another instanceof DoubleClass)
            return DoubleClass.get(number.doubleValue() - another.number.doubleValue());
        else if (this instanceof LongClass || another instanceof LongClass)
            return LongClass.get(number.longValue() - another.number.longValue());

        return IntClass.get(number.intValue() - another.number.intValue());
    }

    public NumberClass multiplication(NumberClass another) {
        if (this instanceof DoubleClass || another instanceof DoubleClass)
            return DoubleClass.get(number.doubleValue() * another.number.doubleValue());
        else if (this instanceof LongClass || another instanceof LongClass)
            return LongClass.get(number.longValue() * another.number.longValue());

        return IntClass.get(number.intValue() * another.number.intValue());
    }

    public NumberClass div(NumberClass another) {
        return DoubleClass.get(number.doubleValue() / another.number.doubleValue());
    }

    public NumberClass pow(NumberClass another) {
        return DoubleClass.get(Math.pow(number.doubleValue(), another.number.doubleValue()));
    }

    public BoolClass larger(NumberClass another) {
        return BoolClass.get(number.doubleValue()>another.number.doubleValue());
    }

    public BoolClass largerEquals(NumberClass another) {
        return BoolClass.get(number.doubleValue()>=another.number.doubleValue());
    }

    public BoolClass less(NumberClass another) {
        return BoolClass.get(number.doubleValue()<another.number.doubleValue());
    }

    public BoolClass lessEquals(NumberClass another) {
        return BoolClass.get(number.doubleValue()<=another.number.doubleValue());
    }
}

File: lclang/src/main/kotlin/lclang/libs/lang/NumberClass.kt
package lclang.libs.lang

import lclang.LCClass
import lclang.method
import lclang.types.Types
import kotlin.math.pow

open class NumberClass(name: String = "number", val number: Number = 0): LCClass(name) {
    init {
        extendsClass = if(name!="number") NumberClass() else null
        globals["toInt"] = method(returnType = Types.INT) { IntClass(number.toInt()) }
        globals["toDouble"] = method(returnType = Types.DOUBLE) { DoubleClass(number.toDouble()) }
        globals["toLong"] = method(returnType = Types.LONG) { LongClass(number.toLong()) }
        globals["toChar"] = method(returnType = Types.CHAR) { CharClass(number.toChar()) }
    }

    operator fun plus(numberClass: NumberClass): NumberClass =
        when{
            this is DoubleClass||numberClass is DoubleClass -> DoubleClass(number.toDouble()+numberClass.number.toDouble())
            this is LongClass||numberClass is LongClass -> LongClass(number.toLong()+numberClass.number.toLong())
            else -> IntClass(number.toInt()+numberClass.number.toInt())
        }

    operator fun minus(numberClass: NumberClass) =
        when{
            this is DoubleClass||numberClass is DoubleClass -> DoubleClass(number.toDouble()-numberClass.number.toDouble())
            this is LongClass||numberClass is LongClass -> LongClass(number.toLong()-numberClass.number.toLong())
            else -> IntClass(number.toInt()-numberClass.number.toInt())
        }

    operator fun times(numberClass: NumberClass) =
        when{
            this is DoubleClass||numberClass is DoubleClass -> DoubleClass(number.toDouble()*numberClass.number.toDouble())
            this is LongClass||numberClass is LongClass -> LongClass(number.toLong()*numberClass.number.toLong())
            else -> IntClass(number.toInt()*numberClass.number.toInt())
        }

    operator fun div(numberClass: NumberClass) = DoubleClass(number.toDouble()/numberClass.number.toDouble())
    fun pow(numberClass: NumberClass) = DoubleClass(number.toDouble().pow(numberClass.number.toDouble()))
    fun larger(numberClass: NumberClass) = BoolClass(number.toDouble() > numberClass.number.toDouble())
    fun largerEquals(numberClass: NumberClass) = BoolClass(number.toDouble() >= numberClass.number.toDouble())
    fun less(numberClass: NumberClass) = BoolClass(number.toDouble() < numberClass.number.toDouble())
    fun lessEquals(numberClass: NumberClass) = BoolClass(number.toDouble() <= numberClass.number.toDouble())
}